AgentOps-AI/agentops
# Pinecone Native Integration

# Pinecone Integration with Comprehensive Testing Suite

## Overview
This PR introduces a comprehensive Pinecone integration along with extensive testing coverage across vector operations, RAG implementations, inference capabilities, and assistant functionalities.

## Key Features

### `PineconeProvider` Class
- Implements a robust provider class that wraps Pinecone's functionality with instrumentation.
- Handles both data plane and control plane operations.
- Includes comprehensive error handling and event tracking.
- Supports both traditional Pinecone operations and newer features like the Assistant API.
- **Note:** Currently, there is no support for evaluating answers using the Assistant API.

### Testing Suites

#### RAG Pipeline Test (`pinecone_rag_test.py`)
- Tests the complete RAG (Retrieval-Augmented Generation) workflow.
- Covers index creation, document embedding, and semantic search.
- Implements vector operations (upsert, update, delete).
- Tests collection management.
- Includes real-world query testing with context retrieval.

#### Inference API Test (`pinecone_inference_test.py`)
- Tests Pinecone's embedding generation capabilities.
- Implements document reranking functionality.
- Validates usage tracking and response formatting.
- Tests various embedding models and configurations.

#### Assistant API Test (`pinecone_assistant_test.py`)
- Tests Pinecone's Assistant API functionality.
- Covers assistant creation, updating, and deletion.
- Implements file upload and management.
- Tests chat completions using an OpenAI-compatible interface.
- **Note:** There is currently no support for evaluating answers within the Assistant API (premium tier feature)

## Implementation Details
- Comprehensive error handling and logging are included.
- Event tracking for all operations is implemented.
- Support for both synchronous and streaming responses is provided.
- Proper cleanup of resources after each test is ensured.
- Integrated with AgentOps for monitoring and tracking.

## Testing Instructions
1. Ensure that the required environment variables are set:
   - `PINECONE_API_KEY`
   - `OPENAI_API_KEY` (for RAG testing)

2. Run individual test suites with the following commands:

    ```bash
    python tests/core_manual_tests/providers/pinecone_rag_test.py
    python tests/core_manual_tests/providers/pinecone_inference_test.py
    python tests/core_manual_tests/providers/pinecone_assistant_test.py
    ```

## Notes
- All tests are designed to be idempotent, ensuring no repeated impact on test results.
- Comprehensive resource cleanup is included in all test flows.
- Tests include rate limiting and appropriate delays to ensure API stability.
- Extensive error handling and detailed reporting are part of the suite.

## Future Improvements
- Add support for batch operations to enhance efficiency.
- Implement retry logic for handling transient failures.
- Integrate custom embedding model support.
- Enhance streaming response handling for smoother real-time interactions.

For more information, refer to the [Pinecone Documentation](https://docs.pinecone.io/reference/api/introduction).
